contract-deconstruct/src/utils.ts

```diff
@@ -0,0 +1,45 @@
+import { TransactionEvent, ethers, getEthersProvider } from "forta-agent";
+
+var ethProvider: ethers.providers.JsonRpcProvider;
+
+export async function getCreatedContractsTX(txEvent: TransactionEvent){
+
+    const contracts: {
+        deployer: string,
+        contractAddress: string,
+        transaction: string,
+    }[] = [];
+
+    let transaction = txEvent.transaction;
+
+    if(transaction.to === null || transaction.to === "0x0000000000000000000000000000000000000000"){
+        if(!ethProvider) ethProvider = getEthersProvider();
+        const receipt = await ethProvider.getTransactionReceipt(transaction.hash);
+        if(!receipt) return;
+        const contractAddressFromReceipt = receipt.contractAddress;
+        contracts.push({
+            deployer: transaction.from,
+            contractAddress: contractAddressFromReceipt,
+            transaction: transaction.hash,
+        })
+    } else {
+        if(!ethProvider) ethProvider = getEthersProvider();
+        const bytecode = await ethProvider.getCode(transaction.to);
```

Comment:
So as I get:
- This bot will be triggered for ANY new transaction.
- If the transaction is not toward the zero address (tx to a proper `to` account) then it will use the Alchemy key.
- What it will do is to check wheter the `to` is a contract or not.
- If it is, it will get traces and check for create/create2 opcodes.

What I see here is that we can "save" some use of the Alchemy premium key (to reduce premium usage). Ideally we should use Alchemy only to retrieve things that the default provider (cloudfare) can't provide -> traces. We have no way to know ahead that we need traces without actually scanning for create/create2 but I think we can save the part where it checks whether the `to` is a contract or not.

Can you find a way to use default cloudfare web3 provider only for this and then IF it is a contract, use Alchemy to scan the traces ?

This can be achievable by not using Alchemy by default in forta.config.json but only on demand once we know that that's a contract to take traces out of it.

I'll leave this here for you to further explore
https://docs.forta.network/en/latest/sensitive-data/

Let me know if this is something achievable